File: src/CombinationSum.java
import java.util.ArrayList;
import java.util.List;

/**
 Given a set of candidate numbers (C) (without duplicates) and a target number (T),
 find all unique combinations in C where the candidate numbers sums to T.

 The same repeated number may be chosen from C unlimited number of times.

 Note:
 All numbers (including target) will be positive integers.
 The solution set must not contain duplicate combinations.
 For example, given candidate set [2, 3, 6, 7] and target 7,
 A solution set is:

 */
public class CombinationSum {
    public static void main(String[] args) {
        int[] candidates = new int[]{ 2, 3, 6, 7 };

        CombinationSum c = new CombinationSum();

        List<List<Integer>> results = c.combinationSum(candidates, 7);

        for (List<Integer> list : results) {
            Printer.printIntegerList(list);
        }
    }

    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> results = new ArrayList<>();

        helper(results, new ArrayList<Integer>(), candidates, target, 0);

        return results;
    }

    public void helper(List<List<Integer>> results,
                       List<Integer> current,
                       int[] candidates,
                       int target,
                       int index) {
        if (target < 0) {
            return;
        }

        if (target == 0) {
            results.add(current);
            return;
        }

        for (int i = index; i < candidates.length; i++) {
            current.add(candidates[i]);

            helper(results,
                    new ArrayList<Integer>(current),
                    candidates,
                    target - candidates[i],
                    i);

            current.remove(current.size() - 1);
        }
    }

}

File: python/Stack/Combination/combination-sum.py
'''

135. Combination Sum
Given a set of candidate numbers (C) and a target number (T),
find all unique combinations in C where the candidate numbers sums
to T.

The same repeated number may be chosen from C unlimited
number of times.

Example
Given candidate set [2,3,6,7] and target 7, a solution set is:

[7]
[2, 2, 3]
Notice
All numbers (including target) will be positive integers.
Elements in a combination (a1, a2, … , ak) must be
in non-descending order. (ie, a1 ≤ a2 ≤ … ≤ ak).
The solution set must not contain duplicate combinations.

'''

def combinationSum(self, candidates, target):
  candidates = list(set(candidates))
  candidates.sort()

  res = []
  self.Dfs(candidates, target, 0, [], res)

def dfs(self, candidates, target, start, curList, res):
  if target == 0:
    res.append(list(curList))
    return

  if target < candidates[start]:
    return

  for i in range(start, len(candidates)):
    curList.append(candidates[i])
    self.dfs(candidates, target - candidate[i], i, curList, res)
    curList.pop()
